### ISSUE TITLE
get feature names from a trained model, python

### ISSUE THREAD

**Issue by Shameendra** [dmlc/xgboost#5275]
Hi i have a pre trained XGBoost CLassifier. I want to find out the name of features/the name of Dataframe columns with which it was trained to i can prepare a table with those features for my use. Can anyone tell me how this can be done. And would be nice if i could get the datatype that the feature expects(eg. int, float or str)

    from xgboost import XGBClassifier
    clf =  XGBClassifier(learning_rate =0.5,
                 n_estimators=207,
                 max_depth=2,
                 min_child_weight=1,
                 gamma=0.8,
                 subsample=0.8,
                 colsample_bytree=0.8,
                 reg_alpha=0.001,
                 objective= 'binary:logistic',
                 nthread=4,
                 scale_pos_weight=1,
                 seed=27)
So how do i get the feature names from this? Any hints would be appriciated

**Comment by trivialfis** [comment]
Could you describe the issue in more details?  There s a feature name property in booster.

**Comment by Shameendra** [comment]
Hi i have a pre trained XGBoost CLassifier. I want to find out the name of features/the name of Dataframe columns with which it was trained to i can prepare a table with those features for my use. Can anyone tell me how this can be done. And would be nice if i could get the datatype that the feature expwects(eg. int, float or str)

    from xgboost import XGBClassifier
    clf =  XGBClassifier(learning_rate =0.5,
                     n_estimators=207,
                     max_depth=2,
                     min_child_weight=1,
                     gamma=0.8,
                     subsample=0.8,
                     colsample_bytree=0.8,
                     reg_alpha=0.001,
                     objective= 'binary:logistic',
                     nthread=4,
                     scale_pos_weight=1,
                     seed=27)


So how do i get the feature names from this? Any hints would be appriciated

**Comment by Shameendra** [comment]
@trivialfis I have updated my question. Sorry i forgot to add details the first time

**Comment by trivialfis** [comment]
No problem.  Call `get_booster()` from `clf`.  Then you have access to the `Booster` object, which contains the `feature_names`.

**Comment by Shameendra** [comment]
@trivialfis It works. Thanks. But it seems that instead of the original names, they have been converted to 'f0','f1','f2' etc. Any way to resture them to their original names?

**Comment by trivialfis** [comment]
Could you provide a self contained script?  With pandas and cuDF, XGBoost should collect the feature names from training matrix.  Also what's your XGBoost version?

**Comment by Shameendra** [comment]
Unfortunately i cannot provide the script scince my company does not allow data shring online . but i can tell you my procedure.

 `
df = pd.read_csv('mycsv.csv', encoding='ISO-8859-1', sep=',')  #read csv
X = df5.drop(['Z'], axis=1)  #create training variables
y = df5['Z'] #create target variables

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20 #Test Train split

X_train_n = X_train.values #convert to np array
X_test_n = X_test.values  #convert to np array

clf = XGBClassifier(learning_rate =0.5,
                     n_estimators=207,
                     max_depth=2,
                     min_child_weight=1,
                     gamma=0.8,
                     subsample=0.8,
                     colsample_bytree=0.8,
                     reg_alpha=0.001,
                     objective= 'binary:logistic',
                     nthread=4,
                     scale_pos_weight=1,
                     seed=27)

clf = xg.get_booster()
print(clf.feature_names)
   
 `

The output i get is 

`
['f0', 'f1', 'f2', 'f3', 'f4', 'f5', 'f6', 'f7', 'f8', 'f9', 'f10', 'f11', 'f12', 'f13', 'f14', 'f15']
`

**Comment by trivialfis** [comment]
Could you try the latest release?  There's nightly build available, see https://xgboost.readthedocs.io/en/latest/build.html  I can't reproduce your error with lastest release.  Also take a look into https://github.com/dmlc/xgboost/issues/5253

``` python
import xgboost as xgb
import numpy as np
import pandas as pd

np.random.seed(1994)

kRows = 64
kCols = 8
kClass = 4

X = np.random.randn(kRows, kCols)
y = np.random.randint(0, kClass, size=kRows)


X = pd.DataFrame(X, columns=['feat-' + str(i) for i in range(kCols)])

cls = xgb.XGBClassifier()
cls.fit(X, y)
print(cls.get_booster().feature_names)

```

```
>>> ['feat-0', 'feat-1', 'feat-2', 'feat-3', 'feat-4', 'feat-5', 'feat-6', 'feat-7']
```

**Comment by trivialfis** [comment]
Or pls make sure your input data frame does have valid `columns` property.

**Comment by Shameendra** [comment]
oke thanks for the suggestions. i will give it a try
